src/view/reviewManager.ts

```diff
@@ -76,8 +82,30 @@ export class ReviewManager implements vscode.DecorationProvider {
 			vscode.commands.executeCommand('vscode.open', vscode.Uri.file(nodePath.resolve(this._repository.path, params.path)), opts);
 		}));
 		this._disposables.push(_repository.onDidRunGitStatus(e => {
-			// todo, validate state only when state changes.
-			this.updateState();
+			const oldHead = this._previousRepositoryState.HEAD;
+			const newHead = this._repository.HEAD;
+			const sameUpstream = !!oldHead.upstream
```

Comment:
This can probably be simplified to 

```
oldHead.upstream && newHead.upstream && oldHead.upstream.name === newHead.upstream.name && oldHead.upstream.remote === newHead.upstream.remote
```

Reply:
I think that misses the cases where both `oldHead.upstream` and `newHead.upstream` are undefined, which should set this to true

Reply:
Make sense.